src/Parallel/Local.hpp

```diff
@@ -21,8 +21,14 @@ template <typename Proxy>
 auto* local(Proxy&& proxy) {
   // It only makes sense to call .ckLocal() on some kinds of proxies
   static_assert(is_chare_proxy<std::decay_t<Proxy>>::value or
-                is_array_element_proxy<std::decay_t<Proxy>>::value);
-  return proxy.ckLocal();
+                is_array_element_proxy<std::decay_t<Proxy>>::value or
+                is_array_proxy<std::decay_t<Proxy>>::value);
+  if constexpr (is_array_proxy<std::decay_t<Proxy>>::value) {
+    // The array case should be a single-element array serving as a singleton
```

Comment:
Is there a way to assert the array indeed has only a single-element here?

Reply:
I don't think so. The docs [here](https://charm.readthedocs.io/en/latest/faq/manual.html#is-there-a-property-similar-to-thisindex-containing-the-chare-arrays-dimensions-or-total-number-of-elements) say that because array chares can have their size changed dynamically, that `any such value would potentially be stale between access and use`.

I can add an `is_singleton<Proxy>` check that checks if the chare type is `Singleton`.

Reply:
I like @kidder's suggestion. This actually will simplify things a lot.

Reply:
I've thought about this a lot and I actually now disagree with @kidder. This function is typically used as:
```
auto& proxy = Parallel::get_parallel_component<ParallelComponent>(cache);
...*Parallel::local(proxy)...;
```

For array elements, it makes sense you want to pass in a proxy to the element like `*Parallel::local(proxy[array_index])`, because you're not calling things on the full array, you're calling them on the elements.

However, for singletons, the user doesn't (and shouldn't) care how the singleton is implemented. So calling this on a singleton proxy should just be `*Parallel::local(proxy)`, not `*Parallel::local(proxy[0])`. The latter puts the burden on the user to remember how singletons are implemented (as single element chare-arrays with an index `int` that is 0) which I don't think we should do.

Thoughts?

Reply:
@knelli2 Yes, you've identified why I went with the current implementation: to hide the details of the singleton implementation inside `local` so that the user doesn't have to worry about it.

But it's true that adding some check that the Proxy is intended to be a singleton would help avoid a potential bug where someone passes in a (non-singleton) array but forgets to index it. I've been away from this for too long to remember if that's easy to do or not 😅